R/list2df_linter.R

```diff
@@ -0,0 +1,64 @@
+#' Recommend direct usage of `data.frame()` to create a data.frame from a list
+#'
+#' [list2DF()] is the preferred way to turn a list of columns into a data.frame.
+#'   Note that it doesn't support recycling; if that's required, use [data.frame()].
+#'
+#' @examples
+#' # will produce lints
+#' lint(
+#'   text = "do.call(cbind.data.frame, x)",
+#'   linters = list2df_linter()
+#' )
+#'
+#' lint(
+#'   text = "do.call('cbind.data.frame', x)",
+#'   linters = list2df_linter()
+#' )
+#'
+#' lint(
+#'   text = "do.call(cbind.data.frame, list(a = 1, b = 1:10))",
+#'   linters = list2df_linter()
+#' )
+#'
+#' # okay
+#' lint(
+#'   text = "list2df(x)",
+#'   linters = list2df_linter()
+#' )
+#'
+#' lint(
+#'   text = "data.frame(list(a = 1, b = 1:10))",
+#'   linters = list2df_linter()
+#' )
+#'
+#' @evalRd rd_tags("list2df_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+list2df_linter <- function() {
+
+  docall_nolambda_xpath <- "
+  //SYMBOL_FUNCTION_CALL[text() = 'do.call']
+    /parent::expr
+    /following-sibling::expr[1][SYMBOL or STR_CONST]
```

Comment:
right now, the linter will land on the first argument, is that the best place? Or should we include the whole `do.call` expression? (not immediately clear to me, though I lean towards the latter)

```
do.call(cbind.data.frame, ...)
        ^ ~ current    |
^         alternative        |
```

Reply:
I agree. I have done this by adding an `xml_find_all(, "./parent")` before returning. Please let me know if this is not the preferred option.